velox/connectors/hive/iceberg/IcebergDataSource.cpp

```diff
@@ -0,0 +1,235 @@
+  // the deleted row numbers relative to the start of the file, and we need to
+  // convert them into row numbers relative to the start of the split.
+  Mutation mutation;
+  mutation.deletedRows = nullptr;
+
+  if (numDeleteRowsInSplit_ > 0) {
+    auto numBytes = bits::nbytes(size);
+    dwio::common::ensureCapacity<int8_t>(deleteBitmap_, numBytes, pool_);
+    std::memset((void*)deleteBitmap_->as<int8_t>(), 0L, numBytes);
+
+    while (deleteFileContentsOffset_ < deleteFileContents_.size()) {
+      auto deleteFile = deleteFileContents_[deleteFileContentsOffset_];
+      VectorPtr& posVector = deleteFile->childAt(1);
+      const int64_t* deleteRows =
+          posVector->as<FlatVector<int64_t>>()->rawValues();
+
+      auto deleteRowsOffsetBefore = deleteRowsOffset_;
+      while (deleteRowsOffset_ < posVector->size() &&
+             deleteRows[deleteRowsOffset_] <
+                 firstRowInSplit_ + readOffset_ + size) {
+        bits::setBit(
+            deleteBitmap_->asMutable<uint64_t>(),
+            deleteRows[deleteRowsOffset_] - readOffset_ - firstRowInSplit_);
+        deleteRowsOffset_++;
+      }
+
+      numDeleteRowsInSplit_ -= deleteRowsOffset_ - deleteRowsOffsetBefore;
+      if (deleteRowsOffset_ == posVector->size()) {
+        deleteFileContentsOffset_++;
+        deleteRowsOffset_ = 0;
+      } else {
+        break;
+      }
+    }
+
+    deleteBitmap_->setSize(numBytes);
+    mutation.deletedRows = deleteBitmap_->as<uint64_t>();
+  }
+
+  auto rowsScanned = HiveDataSource::rowReader_->next(size, output_, &mutation);
+  readOffset_ += rowsScanned;
+
+  return rowsScanned;
+}
+
+void HiveIcebergDataSource::readPositionalDeletes(
+    const IcebergDeleteFile& deleteFile) {
+  auto filePathColumn = ICEBERG_DELETE_FILE_PATH_COLUMN();
+  auto positionsColumn = ICEBERG_DELETE_FILE_POSITIONS_COLUMN();
+
+  std::vector<std::string> deleteColumnNames(
+      {filePathColumn->name, positionsColumn->name});
+  std::vector<std::shared_ptr<const Type>> deleteColumnTypes(
+      {filePathColumn->type, positionsColumn->type});
+
+  RowTypePtr deleteRowType =
+      ROW(std::move(deleteColumnNames), std::move(deleteColumnTypes));
+
+  std::unordered_map<std::string, std::shared_ptr<connector::ColumnHandle>>
+      deleteColumnHandles;
+  deleteColumnHandles[filePathColumn->name] =
+      std::make_shared<HiveColumnHandle>(
+          filePathColumn->name,
+          HiveColumnHandle::ColumnType::kRegular,
+          VARCHAR(),
+          VARCHAR());
+  deleteColumnHandles[positionsColumn->name] =
+      std::make_shared<HiveColumnHandle>(
+          positionsColumn->name,
+          HiveColumnHandle::ColumnType::kRegular,
+          BIGINT(),
+          BIGINT());
+
+  openDeletes(deleteFile, deleteRowType, deleteColumnHandles);
+}
+
+void HiveIcebergDataSource::readEqualityDeletes(
+    const IcebergDeleteFile& deleteFile) {
+  VELOX_NYI();
+}
+
+void HiveIcebergDataSource::openDeletes(
```

Comment:
@Yuhta I intended to do this in future PRs. Do we have to do it now?

Reply:
This will have a rather big impact on how you structure the thing, so better to do it right now, otherwise most of what you write now need to be rewritten.